microservices/update_user_ext_source.py

```diff
@@ -0,0 +1,282 @@
+from adapters.AdaptersManager import AdaptersManager
+from satosa.micro_services.base import ResponseMicroService
+from typing import List, Union, Any
+from utils.ConfigStore import ConfigStore
+
+import threading
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class UpdateUserExtSource(ResponseMicroService):
+    def __init__(self, config, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.DEBUG_PREFIX = "update_user_ext_source: "  # ??
+        self.NAMESPACE = "namespace"
+        self.UES_ATTR_NMS = "urn:perun:ues:attribute-def:def"
+        self.DEFAULT_CONFIG = {
+            'sourceIdPAttributeKey': 'sourceIdPEntityID',
+            'userIdentifiers': ['eduPersonUniqueId',
+                                'eduPersonPrincipalName',
+                                'eduPersonTargetedID',
+                                'nameid',
+                                'uid']
+        }
+
+        self.__adapters_manager = AdaptersManager()
+        self.__attr_map = config['attrMap']
+        self.__append_only_attrs = []
+        self.__array_to_str_conversion = []
+        if config['arrayToStringConversion']:
+            self.__array_to_str_conversion = config['arrayToStringConversion']
+        if config['arrayToStringConversion']:
+            self.__append_only_attrs = config['appendOnlyAttrs']
+
+    def process(self, context, data):
+        data_to_conversion = {
+            'attributes': data['Attributes'],
+            'attrMap': self.__attr_map,
+            'attrsToConversion': self.__array_to_str_conversion,
+            'appendOnlyAttrs': self.__append_only_attrs,
+            'perunUserId': data['perun']['user'].id
+            # Not sure if it contains user data or user instance
+        }
+
+        threading.Thread(
+            target=self.__run,
+            args=(data_to_conversion,)
+        ).start()
+
+        return super().process(context, data)
+
+    def __run(self, data_to_conversion):
+        attrs_from_idp = data_to_conversion['attributes']
+        attr_map = data_to_conversion['attrMap']
+        serialized_attrs = data_to_conversion['attrsToConversion']
+        append_only_attrs = data_to_conversion['appendOnlyAttrs']
+        user_id = data_to_conversion['perunUserId']
+
+        config = self.__get_configuration()
+
+        source_idp_attribute = config['sourceIdPAttributeKey']
+        identifier_attributes = config['userIdentifiers']
+
+        try:
+            if not attrs_from_idp[source_idp_attribute[0]]:
+                # I presume this is where ext source id is stored
+                # in php it worked with ext source name,
+                # but I was told we will store only id and login

```

Comment:
```suggestion
            if not attrs_from_idp[source_idp_attribute][0]:
```
this is the same as `data['attributes'][source_idp_attribute][0]`.

Reply:
or, if you want the name of the IdP = the ext source name, that's `data["auth_info"]["issuer"]`, so here it would be `data_to_conversion["auth_info"]["issuer"]`

Reply:
I guess you want the ext source name directly to eliminate extra call to API.